src/subcommands/mod.rs

```diff
@@ -0,0 +1,54 @@
+// Copyright 2020 Contributors to the Parsec project.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Subcommand implementations. Interacts with parsec-client-rust.
+
+pub mod common;
+pub mod list_opcodes;
+pub mod list_providers;
+pub mod ping;
+
+use crate::cli::ParsecToolApp;
+use crate::error::ParsecToolError;
+use crate::subcommands::{
+    list_opcodes::ListOpcodesSubcommand, list_providers::ListProvidersSubcommand,
+    ping::PingSubcommand,
+};
+use anyhow::Result;
+use parsec_interface::operations::NativeOperation;
+use std::convert::TryInto;
+use structopt::StructOpt;
+
+/// A trait to represent a `parsec-tool` subcommand. Subcommands have three important properties:
+/// - They have their own command-line interface, hence the dependency on `StructOpt`.
+/// - They are convertible to a `NativeOperation` -- i.e. they can all be converted to messages to
+///   the Parsec service. The conversion is fallible.
+/// - They implement `run`, which executes the subcommand.
+pub trait ParsecToolSubcommand: StructOpt + TryInto<NativeOperation> {
+    /// Run the subcommand.
+    fn run(self, matches: ParsecToolApp) -> Result<(), ParsecToolError>;
```

Comment:
I'd probably split that `ParsecToolApp` into `Subcommand` and `Parameters` or `Options` or something like that, where you group all the other parts. And then in this method you pass a reference to that other struct, it feels fishy to give to a subcommand ownership over the structure that presumably holds it? I'm assuming this works because `Subcommand` has `Copy` on it